File: bombenemy.py
class Solution(object):
    def maxKilledEnemies(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        if not grid:
            return 0
        result = 0
        m, n = len(grid), len(grid[0])
        rowhit, colhit = 0, [0]*n
        for i, row in enumerate(grid):
            for j, cell in enumerate(row):
                if j == 0 or grid[i][j-1] == 'W':
                    rowhit = 0
                    k = j
                    while k < n and grid[i][k] != 'W':
                        rowhit += grid[i][k] == 'E'
                        k += 1
                if i == 0 or grid[i-1][j] == 'W':
                    colhit[j] = 0
                    k = i
                    while k < m and grid[k][j] != 'W':
                        colhit[j] += grid[k][j] == 'E'
                        k += 1
                if cell == '0':
                    result = max(result, rowhit+colhit[j])
        
        return result
        

File: bombenemy.cpp
class Solution {
public:
    int maxKilledEnemies(vector<vector<char>>& grid) {
        int m = grid.size();
        if (!m) return 0;
        int n = grid[0].size();
        //vector<vector<int>> dp(m, vector<int>(n, 0));
        int rowhits = 0;
        vector<int> colhits(n, 0);
        int maxVal = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) { 
                if (!j || grid[i][j-1] == 'W') {
                    rowhits = 0;
                    for (int k = j; k < n and grid[i][k]!='W'; k++) rowhits += grid[i][k] == 'E'? 1:0;
                }
                if (!i || grid[i-1][j] == 'W') {
                    colhits[j] = 0;
                    for (int k = i; k < m and grid[k][j] != 'W'; k++) colhits[j] += grid[k][j] == 'E'? 1:0;
                }
                if (grid[i][j] == '0') maxVal = max(maxVal, rowhits+colhits[j]);
            }
        }
        return maxVal;
    }
};
